[HLSL] Implement output parameter

llvm-beanz opened the pull request:
HLSL output parameters are denoted with the `inout` and `out` keywords in the function declaration. When an argument to an output parameter is constructed a temporary value is constructed for the argument.

For `inout` pamameters the argument is intialized by casting the argument expression to the parameter type. For `out` parameters the argument is not initialized before the call.

In both cases on return of the function the temporary value is written back to the argument lvalue expression through an optional casting sequence if required.

This change introduces a new HLSLOutArgExpr ast node which represents the output argument behavior. The OutArgExpr has two defined children: the base expresion and the writeback expression. The writeback expression will either be or contain an OpaqueValueExpr child expression which is used during code generation to represent the temporary value.

Fixes #87526

llvm-beanz committed:
[HLSL] Implement output parameter

HLSL output parameters are denoted with the `inout` and `out` keywords
in the function declaration. When an argument to an output parameter is
constructed a temporary value is constructed for the argument.

For `inout` pamameters the argument is intialized by casting the
argument expression to the parameter type. For `out` parameters the
argument is not initialized before the call.

In both cases on return of the function the temporary value is written
back to the argument lvalue expression through an optional casting
sequence if required.

This change introduces a new HLSLOutArgExpr ast node which represents
the output argument behavior. The OutArgExpr has two defined children:
the base expresion and the writeback expression. The writeback
expression will either be or contain an OpaqueValueExpr child
expression which is used during code generation to represent the
temporary value.

efriedma-quic commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -4057,6 +4057,10 @@ bool Sema::MergeFunctionDecl(FunctionDecl *New, NamedDecl *&OldD, Scope *S,
                                                          NewQType))
       return MergeCompatibleFunctionDecls(New, Old, S, MergeTypeWithOld);
 
+    if (getLangOpts().HLSL && Context.hasSameFunctionTypeIgnoringParamABI(
```
Could you explain a bit more what you're doing here?  At first glance, functions with different out/inout markings aren't compatible.

llvm-beanz replied:
That's correct. We detect and diagnose the mismatch through `mergeParamDeclAttributes`, which gets called by `MergeCompatabileFunctionDecls`, so we need to allow the merge call to proceed even for cases here that might be invalid.

Alternatively we could move that diagnostic out somewhere earlier.

efriedma-quic replied:
Earlier would be more consistent with how other checks like CheckEquivalentExceptionSpec work, I think?

I'm mostly concerned that this logic is twisted in a way that's hard to understand...

llvm-beanz replied:
Can do, will update.

efriedma-quic reviewed (commented):
Approach generally seems fine; not sure I'll have time to look more closely.

llvm-beanz commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -4148,6 +4151,30 @@ static void emitWriteback(CodeGenFunction &CGF,
+      if (TmpVal.isScalar())
+        CGF.EmitStoreThroughLValue(TmpVal, srcLV);
+      else
+        CGF.EmitAggregateStore(srcLV.getPointer(CGF),
```
I think there's a codegen bug (and missing test coverage related to this. I'm debugging right now and will fix and post additional tests once I have it resolved.

llvm-beanz replied:
This should be resolved and covered in the updated tests.

llvm-beanz committed:
Fix aggregate copy emission

This fixes up the code generation errors in non-scalar writeback
operations.

llvm-beanz committed:
Move parameter ABI qualifier mismatch

This change moves the verification for parameter ABI qualifier mismatch
earlier and only merges declarations if the parameter modifiers are
compatible.

There is one case where we do need to merge parameter ABI annotations,
this is in the case for explicit `in` and implicit `in` (no attribute).
All other cases are handled during attribute merging.

This addresses feedback from @efriedma-quic, thank you!

llvm-beanz committed:
Fix clang-format

efriedma-quic commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -4049,6 +4029,13 @@ bool Sema::MergeFunctionDecl(FunctionDecl *New, NamedDecl *&OldD, Scope *S,
+    if (getLangOpts().HLSL) {
+      if (HLSL().CheckCompatibleParameterABI(New, Old))
+        return true;
+      return MergeCompatibleFunctionDecls(New, Old, S, MergeTypeWithOld);
```
`return MergeCompatibleFunctionDecls();` doesn't look right... I suspect this breaks the check for conflicting types.

Maybe we need to change hasSameFunctionTypeIgnoringExceptionSpec to hasSameFunctionTypeIgnoringExceptionSpecAndParameterABI?

llvm-beanz replied:
I wasn't entirely sure about that myself, but couldn't come up with a test case that illustrated a breakage. It is probably a failing of imagination on my part.

`hasSameFunctionTypeIgnoringExceptionSpec` is used in a bunch of places where I _think_ ignoring the parameter ABI would be a problem. Probably not for HLSL because HLSL doesn't support exceptions, but maybe for other cases that use parameter ABIs (swift interop).

I'll put up a patch shortly that brings back the `ASTContext::hasSameFunctionTypeIgnoringParamABI`, and uses that in conjunction with `hasSameFunctionTypeIgnoringExceptionSpec` here to be conservative. Does that sound reasonable?

llvm-beanz committed:
Limit when function declarations are merged

This change makes it so that function declarations are only merged when
no parameter ABI mismatches occur and when the functions have the same
type ignoring the paramater ABIs.

This allows handling cases where implict and explicit parameter ABIs
are the same as with the implicit in and explict in for HLSL.

damyanp commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4613,14 +4613,13 @@ def HLSLGroupSharedAddressSpace : TypeAttr {
 }
 
-def HLSLParamModifier : TypeAttr {
+def HLSLParamModifier : ParameterABIAttr {
```
as an aside: what's the plan for / state of "uniform".  It's mentioned in the docs at least.

llvm-beanz replied:
DXC currently parses but ignores the `uniform` keyword. We have an issue to just officially remove it from the language in 202x since it doesn't do anything, so my plan is to not implement it in Clang at all (https://github.com/microsoft/hlsl-specs/issues/135)

damyanp commented on clang/lib/AST/Expr.cpp:
```diff
@@ -5318,3 +5319,13 @@ OMPIteratorExpr *OMPIteratorExpr::CreateEmpty(const ASTContext &Context,
+HLSLOutArgExpr *HLSLOutArgExpr::Create(const ASTContext &C, QualType Ty,
+                                       Expr *Base, bool IsInOut, Expr *WB,
+                                       OpaqueValueExpr *OpV) {
+  return new (C) HLSLOutArgExpr(Ty, Base, WB, OpV, IsInOut);
```
For my education, how do you decide whether to use `C.Allocate` versus just `C`?

llvm-beanz replied:
Generally `C.Allocate` is used in contexts where the new operator won't know the full size you're allocating. In some AST nodes we allocate extra space after the node to store trailing pointers to sub-expressions. That can be valuable for reducing the size of the AST if a node can have a variable number of sub-expressions.

In this case we always have two sub-expressions, and we always have an `OpaqueValueExpr`. I could reduce the size of the AST node by not caching the `OpaqueValueExpr` and instead traversing down to it when we need it, but I'm unsure that memory savings is worthwhile.

damyanp commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -4689,6 +4719,32 @@ void CodeGenFunction::EmitCallArg(CallArgList &args, const Expr *E,
+    llvm::Value *LifetimeSize = EmitLifetimeStart(Sz, BaseAddr);
+
+    Address TmpAddr(Addr, ElTy, LV.getAlignment());
+    // TODO-HLSLOutArgExp: Fix me!!!
```
Is it more obvious to an expert what needs fixing here?

llvm-beanz replied:
Doh. I think that's just a stale comment, but let me verify that all the test coverage is correct.

damyanp commented on clang/test/SemaHLSL/Language/OutputParameters.hlsl:
```diff
@@ -0,0 +1,34 @@
+
+// Case 1: Calling out and inout parameters with types that cannot be
+// back-converted. In HLSL 2021 and earlier this only occurs when passing scalar
+// arguments to vector parameters because scalar->vector converison is implicit,
```
```suggestion
// arguments to vector parameters because scalar->vector conversion is implicit,
```

damyanp commented on clang/test/SemaHLSL/Language/OutputParameters.hlsl:
```diff
@@ -0,0 +1,34 @@
+  double f;
+  OutFloat(f); // expected-warning{{implicit conversion increases floating-point precision: 'float' to 'double'}}
+  InOutFloat(f); // expected-warning{{implicit conversion increases floating-point precision: 'float' to 'double'}} expected-warning{{implicit conversion loses floating-point precision: 'double' to 'float'}}
+}
```
Are implicit conversations from float -> float1 of any interest here?

llvm-beanz replied:
`case1` above here covers that case. HLSL allows scalar -> vector extension as an implicit cast but not vector -> scalar flattening (which is what float -> float1 is). As such we don't allow vector-scalar mismatches in inout parameters because we need to be able to generate a valid implicit conversion in both directions between the argument and parameter.

llvm-beanz committed:
Resolve FIXME/TODO notes

One of these was left behind to properly put in `noalias` call argument
attributes, the other was there while I was working out the writeback
logic and is resolved and verified by the tests.

llvm-beanz committed:
One last fix to correctly annotate parameter types

We should be annotating the parameter types as restrict references.
This allows the alias information within a function to correctly
identify non-aliasing parameters.

llvm-beanz committed:
Update clang/test/SemaHLSL/Language/OutputParameters.hlsl

Co-authored-by: Damyan Pepper <damyanp@microsoft.com>

llvm-beanz committed:
Fix include ordering

I swear I know my A, B, C's...

llvm-beanz commented:
@rjmccall curious if you have some time to look at this PR. I've made some small changes to extend parameter ABI annotations and write back arguments (used by ObjC) to cover HLSL's odd parameter behaviors.

coopp reviewed (approved):
That was a lot!.  Didn't see anything to call out, so it looks good to me.

rjmccall commented on clang/include/clang/AST/Expr.h:
```diff
@@ -7061,6 +7061,67 @@ class ArraySectionExpr : public Expr {
+                                bool IsInOut, Expr *WB, OpaqueValueExpr *OpV);
+  static HLSLOutArgExpr *CreateEmpty(const ASTContext &Ctx);
+
+  const Expr *getBase() const { return Base; }
```
You need to document what the expressions are here, because it's pretty subtle.  So the way you've written it, this a conversion of the base l-value to the type of the parameter, the writeback is an assignment of the temporary result back to the base l-value, and the OVE represents the temporary result.  I think this is problematic, though; see the comment in Sema.

rjmccall commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -5915,6 +5916,16 @@ bool Sema::GatherArgumentsForCall(SourceLocation CallLoc, FunctionDecl *FDecl,
         if (auto *BE = dyn_cast<BlockExpr>(Arg->IgnoreParenNoopCasts(Context)))
           BE->getBlockDecl()->setDoesNotEscape();
+      // During template instantiation it is possible that we already created
+      // the HLSLOutArgExpr if it was produced during tree transformation.
```
Generally, `TreeTransform` just drops these implicit nodes on the floor, instantiates the original expression, and creates new nodes when it's re-processing the instantiated call.  This can get complicated around OVEs for e.g. pseudo-object expressions, but in your case, it should be easy because *all* of the extra structure is implicit stuff, and we haven't restructured anything.

rjmccall replied:
Since you've fixed TreeTransform to look through the `HLSLOutArgExpr`, it should now no longer be possible to see an `HLSLOutArgExpr` here.

rjmccall commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -1121,3 +1121,99 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
+  }
+
+  bool RequiresConversion =
+      Ty.getUnqualifiedType() != Arg->getType().getUnqualifiedType();
```
This is actually doing a non-canonical comparison, which is not what you want; the test case is just writing one of these with a `typedef` of the type of the other.  There's a `ASTContext::hasSameUnqualifiedType` that you should use.

rjmccall commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -1121,3 +1121,99 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
+  // results in a copy we can ignore qualifiers.
+  if (RequiresConversion) {
+    ExprResult Res =
+        SemaRef.PerformImplicitConversion(Arg, Ty, Sema::AA_Passing);
```
Okay.  It looks like you're losing track of `Arg` here.  I honestly don't understand how CodeGen works in the converted case; it seems like it'd have to be wrong, because you're calling `EmitLValue` on `Base` when it isn't even an l-value anymore.

The most obvious representation would be something like this:
- Make an OVE for the original l-value (`Arg`).  You can use `Arg` as the source expression.  This is always an l-value.
- Make a conversion from that OVE to the parameter type.  This is always an r-value.
- Make an OVE for the temporary value; this is always an l-value and should not have a source expression.
- Make a conversion from that OVE to the original l-value type.  This is always an r-value.

You could potentially try to recover the original l-value by looking through the implicit conversions on the converted r-value expression.  In C use cases, that might be reliable; you could assert here that your algorithm works.  That would be simpler for the representation, but probably significantly harder in CodeGen without double-emitting the original l-value.

HLSL is a C extension and doesn't support C++, right?  So you don't need to worry about user-defined assignment operators for the writeback?

llvm-beanz replied:
HLSL does support some C++ features. It doesn't (yet) support user-defined assignment operators, casting operators, or constructors, but those are all highly requested features.

llvm-beanz replied:
My latest update is mostly in line with the feedback here. I did the OVE's a little different, but it _radically_ simplified the CodeGen changes (thank you for all the suggestions!).

I've changed the initialization to use `PerformCopyInitialization` so that it will resolve constructors correctly and the write back is just a binary op built with `ActOnBinOp`. That also radically simplified the codegen.

At a simplified level the C++-equivalent code for how `inout` and `out` arguments work is something like this:

```c++
namespace detail {
template<typename TmpT, typename SrcT>
struct InOutAdapter {
    SrcT &SrcLV;
    TmpT TmpLV;

    InOutAdapter(SrcT &S) : SrcLV(S), TmpLV(S) {}
    ~InOutAdapter() { SrcLV = TmpLV; }

    operator TmpT&(){
        return TmpLV;
    }
};
}

template <typename ArgT, typename SrcT>
detail::InOutAdapter<ArgT, SrcT> make_inout(SrcT &V) {
    return detail::InOutAdapter<ArgT, SrcT>(V);
}

namespace detail {
template<typename TmpT, typename SrcT>
struct OutAdapter {
    SrcT &SrcLV;
    TmpT TmpLV;

    OutAdapter(SrcT &S) : SrcLV(S) {}
    ~OutAdapter() { SrcLV = TmpLV; }

    operator TmpT&(){
        return TmpLV;
    }
};
}

template <typename ArgT, typename SrcT>
detail::OutAdapter<ArgT, SrcT> make_out(SrcT &V) {
    return detail::OutAdapter<ArgT, SrcT>(V);
}
```
[Compiler Explorer C++ Example](https://godbolt.org/z/G1jhn78qG)

I believe with my latest changes the AST generation and codegen both correctly match this.

llvm-beanz committed:
Merge remote-tracking branch 'origin/main' into cbieneman/inout

llvm-beanz committed:
Update for upstream changes

(more changes coming)

llvm-beanz commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -4148,6 +4152,30 @@ static void emitWriteback(CodeGenFunction &CGF,
+      if (TmpVal.isScalar())
+        CGF.EmitStoreThroughLValue(TmpVal, srcLV);
+      else {
+        llvm::Value *Val = CGF.Builder.CreateLoad(TmpVal.getAggregateAddress());
```
There's a really awful type of lvalue cast that HLSL supports (which I haven't yet added to Clang). We allow something like this:

```hlsl
struct Dog {
  int legs[4];
};

struct Cat {
  int4 legs;
};

void pet(inout Cat C) {
  C.legs.x = 1;
}

export void call() {
  Dog d;
  pet((Cat)d);
}
```

In HLSL 2018, this conversion was allowed to be implicit, but we're slowly tightening up the rules.

I will pull this code out from this PR though, because it isn't possible yet to construct an expression that results in that write back formation.

rjmccall replied:
Well, that's awful, but it's supportable.  The easiest way to get out of the business of having to handle all these different evaluation kinds is to just change the AST to the representation you'll need to use to support non-POD types, because then the load/assignment will just be the normal code-gen for emitting assignments.

llvm-beanz committed:
Update based on PR feedback

This reworks the PR based on feedback from @rjmcall.

There are a few differences here from his suggestions, but I think this
is the right direction.

The AST node now always creates two opaque values. One that has a
source expression of the original argument lvalue, and one that has a
source expression of the parameter temporary.

The casted temporary is initiailzed via copy initialization, and the
writeback follows the same flow. I had some trouble getting ActOnBinOp
working, and while I haven't given up on it I think this approach works
too.

I've added expanded test coverage of side-effecting lvalues,
order-dependent writebacks, writebacks through bitfields, and more.

llvm-beanz committed:
Fixed ActOnBinOp

llvm-beanz committed:
Merge remote-tracking branch 'origin/main' into cbieneman/inout

# Conflicts:
#	clang/include/clang/Sema/SemaHLSL.h
#	clang/lib/Sema/SemaHLSL.cpp

rjmccall commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8511,7 +8514,7 @@ static void HandleHLSLParamModifierAttr(QualType &CurType,
   if (Attr.getSemanticSpelling() == HLSLParamModifierAttr::Keyword_inout ||
       Attr.getSemanticSpelling() == HLSLParamModifierAttr::Keyword_out)
-    CurType = S.getASTContext().getLValueReferenceType(CurType);
+    CurType = S.HLSL().getInoutParameterType(CurType);
```
There are semantic restrictions here, right?  If nothing else, `CurType` can't be a reference type.

llvm-beanz replied:
HLSL doesn't allow references or pointers (at least not explicitly). We generate references in the AST in a few places, but any attempt to write or type-deduce one results in an error in Sema.

In theory, any legal type in HLSL can be converted to an inout parameter type.

rjmccall replied:
You can't reseat a reference, so I don't know what it would mean to pass one `inout`.  It should probably be invalid.

At any rate, if HLSL doesn't allow references, that's fine; please document that expectation with a comment, though (and maybe an assertion that we never see a reference here?).

llvm-beanz replied:
I put an assert in `SemaHLSL::getInoutParameterType` so that it will fire anywhere we try to adjust a type to the inout/out argument type.

rjmccall reviewed (commented):
Thanks, this is looking a lot better.

llvm-beanz commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -11427,6 +11427,19 @@ static void AnalyzeImplicitConversions(
+  if (auto *OutArgE = dyn_cast<HLSLOutArgExpr>(E)) {
+    // The base expression is only used to initialize the parameter for
+    // arguments to `inout` parameters, so we only traverse down the base
+    // expression for `inout` cases.
```
I should have commented on this. If SemaChecking traverses down both sides we get warnings for the conversion that would "initialize" the `out` parameter, but those warnings aren't accurate because no initialization occurs.

If we only traverse down the write back side we only get the warning for the copy back to the l-value, which is the behavior we want.

clang/test/SemaHLSL/Language/OutputParameters.hlsl verifies the behavior for `out` arguments of only diagnosing the write back conversion, removing the condition breaks that test by emitting diagnostics on the initialization as well.

rjmccall replied:
I specifically said the original l-value expression.  I'm trying to make sure we get warnings in sub-expressions of the l-value that have nothing to do with the final conversion to produce the argument.

With that said:
> If SemaChecking traverses down both sides we get warnings for the conversion that would "initialize" the out parameter, but those warnings aren't accurate because no initialization occurs.

Isn't this an argument that you either (1) shouldn't visit the initialization specifically for `out`, but should still do it for `inout`, or (2) shouldn't generate the initialization expression for `out`?

Apropos, I know this is a future direction, but what semantics do you want for C++ class types and `out`?  Skipping the initialization entirely is definitely incorrect — the callee will get passed a reference to an uninitialized object and (presumably) start calling methods on it.  Probably this should default-initialize?

llvm-beanz replied:
Gotcha! I think I misunderstood your comment since I didn't have the original lvalue preserved initially. I'll update this.

These are pretty wonky when we start looking at C++isms. My team reacted with disgust when I posed this potential (but currently invalid) code:

```hlsl
RWBuffer<int> Val;

struct A {
  int Idx;
  A &operator ++() {
    Val[Idx] = Val[Idx] + 1;
    return *this;
  }
};

void init(out A Obj) {
  Obj.Idx = 2;
};

[numthreads(1,1,1)]
void main() {
  A a;
  a.Idx = 0;
  init(++a);
  ++a;
}
```

We certainly need to evaluate the argument expression (so that it can side-effect). Constructors would make this more complicated, but it probably should default initialize. Are you okay with that being a follow-up? If so I'll file an issue to track doing that work.

rjmccall replied:
Yeah, since you're not supporting C++ types here, it's fine to leave the work for later.

llvm-beanz committed:
Updates based on latest feedback

Also got a few things I missed in my last round of updates.

llvm-beanz committed:
clang-format

llvm-beanz committed:
clang-format

llvm-beanz committed:
Merge remote-tracking branch 'origin/main' into cbieneman/inout

llvm-beanz committed:
Add an assert in SemaHLSL::getInoutParameterType

llvm-beanz committed:
Process the initial lvalue in SemaChecking

This processes just the initial lvalue during sema checking since that
expression is always evaluated.

The updated test case has an implicit conversion in the lvalue
expression, which is evaluated but not used to initialize the argument.

llvm-beanz committed:
Avoid double mutating parameter modifiers

This updates the code in SemaHLSL::getInoutParameterType to assert on
reference types, and fixes the cooresponding code in SemaType so that
the underlying type isn't mutated more than once.

The double-mutation only occured in error cases where conflicting
attributes were applied (error is checked elsewhere).

llvm-beanz commented:
ping @rjmccall. I think I've updated to address all your feedback. Let me know if there's anything else you think I should change.

rjmccall commented on clang/include/clang/AST/Expr.h:
```diff
@@ -7071,6 +7071,102 @@ class ArraySectionExpr : public Expr {
+///
+/// This AST node has three sub-expressions:
+///  - An OpaqueValueExpr with a source that is the argument lvalue expression.
+///  - An OpaqueValueExpr with a source that is a an implicit conversion
```
```suggestion
///  - An OpaqueValueExpr with a source that is an implicit conversion
```

Also, you should mention that this isn't evaluated for an `out` parameter.

rjmccall commented on clang/include/clang/AST/Expr.h:
```diff
@@ -7071,6 +7071,102 @@ class ArraySectionExpr : public Expr {
+///    sequence from the source lvalue to the argument type.
+///  - A BinaryOperatorExpr that assigns the first sub-expression with the
+///    value from an implict conversion sequence from the second expression to
+///    the argument expression's type.
```
Maybe just "An expression that assigns the second expression into the first, performing any necessary conversions."

rjmccall commented on clang/include/clang/AST/Expr.h:
```diff
@@ -7071,6 +7071,102 @@ class ArraySectionExpr : public Expr {
+    return cast<OpaqueValueExpr>(SubExprs[BaseLValue]);
+  }
+
+  const Expr *getArgLValue() const {
```
```suggestion
  /// Return the l-value expression that was written as the argument
  /// in source.  Everything else here is implicitly generated.
  const Expr *getArgLValue() const {
```

People writing source tools will appreciate this being explicitly called out.

rjmccall commented on clang/lib/CodeGen/CGCall.h:
```diff
@@ -285,6 +285,13 @@ class CallArgList : public SmallVector<CallArg, 8> {
+
+    /// An Expression representing a cast from the temporary's type to the
+    /// source l-value's type.
+    const Expr *CastExpr;
```
This comment is no longer correct; consider something like `WritebackExpr`. And maybe the logic for writebacks should simply be that we evaluate this expression if it's present, rather than making a language-mode decision?

rjmccall commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -5432,6 +5434,57 @@ LValue CodeGenFunction::EmitOpaqueValueLValue(const OpaqueValueExpr *e) {
+
+  LValue TempLV;
+
+  if (E->isInOut()) {
```
I was hoping you'd extract that as a helper function you'd just call rather than duplicating the logic here.

llvm-beanz committed:
Update clang/include/clang/AST/Expr.h

Co-authored-by: John McCall <rjmccall@gmail.com>

llvm-beanz committed:
Update clang/include/clang/AST/Expr.h

Co-authored-by: John McCall <rjmccall@gmail.com>

llvm-beanz committed:
Update code comment

llvm-beanz committed:
Remove unreachable case

llvm-beanz committed:
Rename CastExpr to WritebackExpr

llvm-beanz committed:
Extract out common helper code

rjmccall reviewed (approved):
Thanks, LGTM.

llvm-beanz committed:
Merge remote-tracking branch 'origin/main' into cbieneman/inout

llvm-beanz committed:
Add comment

(apparently I forgot to save this file)

llvm-beanz committed:
Merge remote-tracking branch 'origin/main' into cbieneman/inout
